clang/lib/Sema/SemaType.cpp

```diff
@@ -3605,8 +3605,11 @@ static QualType GetDeclSpecTypeForDeclarator(TypeProcessingState &state,
         Info = &SemaRef.InventedParameterInfos.back();
       } else {
         // In C++14, generic lambdas allow 'auto' in their parameters.
-        if (!SemaRef.getLangOpts().CPlusPlus14 || !Auto ||
-            Auto->getKeyword() != AutoTypeKeyword::Auto) {
+        if (!SemaRef.getLangOpts().CPlusPlus14 && Auto &&
+            Auto->getKeyword() == AutoTypeKeyword::Auto) {
+          Error = 24;
```

Comment:
A lot of the other cases have comments explaining what diagnostic the code pertains to. See 20 or 7 below for example. 

This is not really great but at least we have a guide to which diagnostic we intend to emit.

Reply:
Makes sense. I added some similar comments as well.